### PR TITLE
Handle hash patterns and pins in `Lint/OutOfRangeRegexpRef` cop

### PR DESCRIPTION
**Pull request by fatkodima**
Follow up to https://github.com/rubocop/rubocop/pull/11337#discussion_r1103995680

### CODE REVIEW

File: spec/rubocop/cop/lint/out_of_range_regexp_ref_spec.rb [hunk 1]
```diff
@@ -234,6 +234,46 @@
...
+        RUBY
+
+        expect_offense(<<~RUBY)
+          case hash
+          in a: /(foo)(bar)/, b: /(bar)baz/
+            $3
+            ^^ $3 is out of range (2 regexp capture groups detected).
+          end
+        RUBY
+      end
```
Comment (koic) [inline spec/rubocop/cop/lint/out_of_range_regexp_ref_spec.rb:253]: Can you make `expect_no_offenses` and `expect_offense` different `it` for test case isolation?

File: spec/rubocop/cop/lint/out_of_range_regexp_ref_spec.rb [hunk 2]
```diff
@@ -234,6 +234,46 @@
...
+
+        expect_offense(<<~RUBY)
+          a = 1
+          case array
+          in [^a, /(foo)(bar)/, /(foo)bar/]
+            $3
+            ^^ $3 is out of range (2 regexp capture groups detected).
+          end
+        RUBY
+      end
```
Comment (koic) [inline spec/rubocop/cop/lint/out_of_range_regexp_ref_spec.rb:274]: Ditto.

**Comment by fatkodima** [comment]
Updated test cases.

**Comment by koic** [comment]
Thanks!

**Pull request merged**